shadowsocks/tcp.go

```diff
@@ -231,6 +241,16 @@ func (s *tcpService) Start() {
 	}
 }
 
+// Keep the connection open until we hit the authentication deadline to protect against probing attacks
+// `proxyMetrics` is a pointer because its value is being mutated by `clientConn`.
+func (s *tcpService) absorbProbe(clientConn io.ReadCloser, clientLocation string, proxyMetrics *metrics.ProxyMetrics) {
+	_, drainErr := io.Copy(ioutil.Discard, clientConn) // drain socket
+	drainResult := drainErrToString(drainErr)
+	port := s.listener.Addr().(*net.TCPAddr).Port
+	logger.Debugf("Drain error: %v, drain result: %v", drainErr, drainResult)
+	s.m.AddTCPProbe(clientLocation, drainResult, port, *proxyMetrics)
```

Comment:
We should change this to indicate whether the probe is a failed auth or replay.

Reply:
Done